### PR TITLE
New Visualisation: Call graph

### PR DESCRIPTION
**Pull request by todd-dembrey**
Adds a new visualisation, see below for an example. The boxes are colored based on time in the functions (red=more, green = less).

There is a reasonable amount of refactoring within the JavaScript that happened alongside this development. If the visualisation is not acceptable perhaps that can be a separate pull request.

callgraph

### CODE REVIEW

**Comment by jiffyclub** [comment]
Thanks for the PR, hopefully I'll get a chance to review it soon. It looks like this includes #64? One thing: I'd like to keep the indentation consistent at 2 spaces.

**Comment by todd-dembrey** [comment]
Yes it does include #64 (should have kept independence probably)

**Comment by earthgecko** [comment]
@todd-dembrey that is another nice addition, helps to see the main hotspots in a different kind of visualisation than those provided by Sunburst and Icicle.

hotspots

SnakeViz and Sunburst seems a much more intuitive visualisation than runsnakerun, thanks @jiffyclub looks d3js-esque, surprised to see that it was not.  I suspect that SnakeViz with d3js could produce some wonderful art :)

File: snakeviz/static/snakeviz.js [hunk 1]
```diff
@@ -1,227 +1,278 @@
...
+    callStack.checkResetButtonState();
+  }
+  var message = buildMessage(callStack.currentRoot(),callStack.currentParent(),init);
+  sv_worker.postMessage(message);
+  _.defer(sv_hide_working);
 };
 
+var buildMessage = function(rootName, parentName,initialRun){
+	if (initialRun){
+		depth = 10000;
```
Comment (jiffyclub) [inline snakeviz/static/snakeviz.js]: What the reasoning behind this initial 10000 / 0? That seems likely to lead to some crashes on large profiles.

File: snakeviz/templates/viz.html [hunk 2]
```diff
@@ -21,8 +21,9 @@ <h1 id="snakeviz-text">SnakeViz</h1>
     <!-- style select -->
     <label id='sv-style-label'>Style:
       <select name="sv-style" id="sv-style-select">
-        <option value="sunburst" selected>Sunburst</option>
+        <option value="sunburst">Sunburst</option>
         <option value="icicle">Icicle</option>
+        <option value="callgraph" selected>Call Graph</option>
```
Comment (jiffyclub) [inline snakeviz/templates/viz.html]: I'd rather not change the default viz.

**Comment by jiffyclub** [comment]
I pulled this down locally and can't seem to get it to work. I'm on a Mac and it looks like this in both Chrome and Firefox:

screenshot 2016-06-11 22 29 54

I tried changing the visualization type (via the source code) but the same things happens. There are no errors in the Chrome console, but the Firefox console says this:

```
Attempting to create a Worker from an empty source. This is probably unintentional.
```

I don't see that error if I run the latest release of SnakeViz. If I inspect the rendered HTML there isn't any SVG data in there (e.g. the `#chart` div is empty).

**Comment by jiffyclub** [comment]
I pulled this down locally and can't seem to get it to work. I'm on a Mac and it looks like this in both Chrome and Firefox:

screenshot 2016-06-11 22 29 54

I tried changing the visualization type (via the source code) but the same things happens. There are no errors in the Chrome console, but the Firefox console says this:

```
Attempting to create a Worker from an empty source. This is probably unintentional.
```

I don't see that error if I run the latest release of SnakeViz. If I inspect the rendered HTML there isn't any SVG data in there (e.g. the `#chart` div is empty).

**Comment by jiffyclub** [comment]
I pulled this down locally and can't seem to get it to work. I'm on a Mac and it looks like this in both Chrome and Firefox:

screenshot 2016-06-11 22 29 54

I tried changing the visualization type (via the source code) but the same things happens. There are no errors in the Chrome console, but the Firefox console says this:

```
Attempting to create a Worker from an empty source. This is probably unintentional.
```

I don't see that error if I run the latest release of SnakeViz. If I inspect the rendered HTML there isn't any SVG data in there (e.g. the `#chart` div is empty).

**Comment by jiffyclub** [comment]
@embray if you have time would you mind also going over this PR?

**Comment by jiffyclub** [comment]
@todd-dembrey I checked out your branch directly and it's working, so I think the problem I had above is related to merging your branch with master.

**Comment by jiffyclub** [comment]
A little feedback now that I've played with a working version of this:

- First of all, this is pretty neat! Thanks for taking the time to work on SnakeViz.
- The sunburst and icicle plots can be constrained to fit the allowed vertical space, but the call graph can overflow, making parts of the graph unreachable:
    screenshot 2016-06-12 08 53 31
- This definitely breaks on large/complicated profiles, and when it breaks the controls are covered by the table so you can't adjust any of the settings and there's no error message:
    screenshot 2016-06-12 09 16 04

If you need a large/complicated profile to work with, these should do the trick: 
[profiles.zip](https://github.com/jiffyclub/snakeviz/files/310849/profiles.zip) The kdv profile works with Python 3 and the skimage profile works with Python 2.

**Pull request closed without merging**